design/ceph/object/store.md

```diff
@@ -712,3 +712,64 @@ For external CephCephObjectStores (i.e., when `spec.gateway.externalRgwEndpoints
 vhost-style addressing should be configured on the host cluster, and `hosting.dnsNames` is
 irrelevant. The default `advertiseEndpoint` for external CephObjectStores is the first entry in the
 `spec.gateway.externalRgwEndpoints` list, which users should be able to override if desired.
+
+### Advanced user configuration
+
+RGW is a very active project with many existing features and regular new feature additions. Rook
+maintainers have a hard time catching up with RGW's latest features. In an effort to help users who
+want access to rare and new features, Rook can allow users to set arbitrary Ceph configs in their
+CephObjectStore CR. RGW config reference: https://docs.ceph.com/en/reef/radosgw/config-ref/
+
+This is an advanced feature. It allows users to modify many untested aspects of the RGW and object
+store. Any user configs that break the object store should be a responsibility of the user.
+
+```yaml
+spec:
+  metadataPool: # ...
+  dataPool: # ...
+  gateway: # ...
+
+  cephConfig:
+    rgw_enable_apis: s3, s3website, swift, swift_auth, sts, iam, notifications # e.g. 'admin' removed
+    rgw_gc_max_concurrent_io: "4" # e.g, reduce garbage collection IO
+    debug_rgw: "20"
+```
+
+CephCluster has a similar configuration section called `config`. `config` may be somewhat too vague
+in the context of the CephObjectStore CR. `cephConfig` is a clearer naming for users and give
+context that Ceph configs are expected.
+
+Rook will apply user config values to the mon config store. Rook will not override its own mon
+config values set via other means. This will ensure that values controlled by Rook for other spec
+configurations or that are important for running in Kubernetes are not clobbered by user configs.
+This will risk some instances where users want to update a value that Rook currently sets.
+This tradeoff favors reliability over flexibility.
+
+Examples:
+    - Rook sets configs like `rgw_zone` which are critical for a functional CephObjectStore. Users
+    would not be able to override this value.
+    - Rook sets `rgw_dns_name` via the `hosting` config section. User inputs to `rgw_dns_name`
+    wouldn't take affect when `hosting` configs are set.
```

Comment:
@travisn what are your thoughts about this design tradeoff?

Reply:
If there are other settings in the CRD that explicitly offer the feature such as for the `hosting`, they should use those rather than these `cephConfig` options. This seems like the same problem we have with `cephConfig` in the CephCluster CR where the user could create conflicting settings. It's an advanced feature that gives all the power, even though we know it can lead to conflicting settings. The user has to manage that complexity themselves with these advanced features.

Reply:
Right. The intent of the design as currently written is to sort of force users to do the thing we want them to do. If Rook is already setting a mon KV config, Rook won't apply the `cephConfig` setting (and it'll log a note for users about it).

Are you in favor of keeping this behavior? Or are you suggesting to change it so these are overrides regardless of what Rook is doing?

Separately, we could have a blocklist of configs that we don't allow users to set themselves (like `rgw_dns_name`, `rgw_zone`, etc. I thought that we agreed not to do that in CephCluster, so I opted not to plan for that here. LMK if you think we should plan for blocklisting here. Also, we can always make this decision in the future. IMO, it isn't critical to the MVP.

Reply:
I would vote to keep this feature as simple as possible. It's advanced, and the user may shoot themselves in the foot. I just don't think it's worth the effort to prevent the user from applying settings.

Reply:
The code effort is the same either way. It's really about order of operations.

Either Rook applies its own preferred settings and the settings applied from specific API inputs first, and then puts the `cephConfig` settings on top of those -- or Rook does it the other way around, applying preferred/specific settings on top of `cephConfig`. 

IMO, the current plan (to overlay any preferred/specific configs on top) is better for ensuring users aren't breaking random stuff. The downside is that inevitably, some user will want to change some value that Rook currently sets as preferred.

Reply:
After the discussion in huddle, I think (1) I'll try to be more clear about this behavior in an updated design, and (2) it sounds like we want to allow users to override Rook's options via this mechanism, with a documented warning that if they break it, they fix it.